lib/main/itoSession.ts

```diff
@@ -0,0 +1,215 @@
+  private streamResponsePromise: Promise<{
+    response: any
+    audioBuffer: Buffer
+    sampleRate: number
+  }> | null = null
+
+  public async startSession(mode: ItoMode) {
+    log.info('[ItoSession] Starting session with mode:', mode)
+
+    // Initialize all necessary components
+    const started = await itoStreamController.startInteraction(mode)
+    if (!started) {
+      log.error('[ItoSession] Failed to start itoStreamController')
+      return
+    }
+
+    // Begin gRPC stream immediately (note, no audio is flowing yet)
+    this.streamResponsePromise = itoStreamController.startGrpcStream()
+
+    // Begin recording audio (audio bytes will now flow into the gRPC stream)
+    voiceInputService.startAudioRecording()
+
+    // Send initial mode to the stream
+    itoStreamController.setMode(mode)
+
+    // Update UI state
+    recordingStateNotifier.notifyRecordingStarted(mode)
+
+    // Fetch and send context in the background (non-blocking)
+    this.fetchAndSendContext().catch(error => {
+      log.error('[ItoSession] Failed to fetch/send context:', error)
+    })
+  }
+
+  private async fetchAndSendContext() {
+    log.info('[ItoSession] Fetching context in background...')
+
+    // This builds the full config (window context, selected text, vocabulary, settings)
+    await itoStreamController.sendConfigUpdate()
+
+    log.info('[ItoSession] Context sent to stream')
+
+    // Fetch cursor context for grammar rules only if grammar service is enabled
+    const { grammarServiceEnabled } = getAdvancedSettings()
+    if (grammarServiceEnabled) {
+      const cursorContext = await contextGrabber.getCursorContextForGrammar(4)
+      grammarRulesService.setCursorContext(cursorContext)
+      log.info('[ItoSession] Cursor context set for grammar rules')
+    }
+  }
+
+  public setMode(mode: ItoMode) {
+    log.info('[ItoSession] Changing mode to:', mode)
+
+    // Send mode change to grpc stream (will also update windows via recordingStateNotifier)
+    itoStreamController.setMode(mode)
+
+    // Update UI to show the new mode
+    recordingStateNotifier.notifyRecordingStarted(mode)
+  }
+
+  public async cancelSession() {
+    log.info('[ItoSession] Cancelling session')
+
+    // Cancel the transcription (will not create interaction)
+    itoStreamController.cancelTranscription()
+
+    // Stop audio recording
+    await voiceInputService.stopAudioRecording()
+
+    // Update UI state
+    recordingStateNotifier.notifyRecordingStopped()
+
+    // Clear cursor context on cancel
+    grammarRulesService.clearCursorContext()
+
+    log.info('[ItoSession] Session cancelled')
+  }
+
+  public async completeSession() {
+    // Stop audio recording and wait for drain
+    await voiceInputService.stopAudioRecording()
+
+    // Check actual audio duration (keyboard duration can be misleading due to latency)
+    const audioDurationMs = itoStreamController.getAudioDurationMs()
+
+    if (audioDurationMs < this.MINIMUM_AUDIO_DURATION_MS) {
```

Comment:
I think this is cleaner to have the itoSession itself cancel its own stream instead of buffering audio and only beginning streaming once we have met the minimum buffer.